### ISSUE TITLE
Feature: LUA callin to allow overriding build position during blueprint preview

### ISSUE THREAD

**Issue by thehobojoe** [beyond-all-reason/spring#1188]
The reasoning here is simple, sometimes widgets will want to change build position from the cursor to somewhere else in the map. For example, BAR's mex snap:

![spring_XlNucfgWQX](https://github.com/beyond-all-reason/spring/assets/3493638/b3d6c6a4-f020-4659-a6a8-b504d4083dab)

The build preview here is not actually where the build order will be given - the build order will be placed at the secondary ghost that is on the mex spot. But, if masks are used to limit the mex placement to mex spots only, they will display as invalid, even though via game-side logic the placement will work. 

![image](https://github.com/beyond-all-reason/spring/assets/3493638/a41e9990-67ab-45c6-ab83-51a7f51785f2)

@sprunk proposed a callin, such as:
```lua
widget:BuildPosition(pos) -- { x, y, z }
    -- calculate new position
    return position -- { x, y, z }
end
```

This should allow for much more robust control of building placement on the game side via widget, which will open up lots of new possibilities.

**Comment by sprunk** [comment]
Sounds great for all kinds of snapping.

Having the `y` param would also make it possible to reflect expected terraform spires (where now it stays at the bottom), though for convenience it would be good if you could keep it `nil`. Perhaps it should be `x, z, y`?
![image](https://github.com/beyond-all-reason/spring/assets/2573076/2a0194e2-6db9-40b7-a888-8cce071e3aee)
![image](https://github.com/beyond-all-reason/spring/assets/2573076/7914563b-a807-4ccd-975f-fe0f53c559d1)

**Comment by lhog** [comment]
I can expect a lot of confusion, especially if the player's monitor is big or they're zoomed in. Like: "wtf where is my blueprint?"

**Comment by sprunk** [comment]
Yes, but it's the responsibility of the game to project it to a sane place or provide some clue. Like existing mex snap just draws a line pointing to the projected spot.

**Comment by thehobojoe** [comment]
Same could be said about nearly any of the LUA API's. Like "wtf why are my units getting commands I didn't give them". It's up to the game to use these tools well, whether they could be confusing if misused is not really a consideration that gets applied to anything else.
